openedx/core/lib/request_utils.py

```diff
@@ -227,3 +229,249 @@ def set_custom_attributes_for_top_n(self, names_to_size, top_n_captured, attribu
             set_custom_attribute(name_attribute, name)
             set_custom_attribute(size_attribute, size)
             log.debug('%s = %d', name, size)
+
+
+def expected_error_exception_handler(exc, context):
+    """
+    Replacement for DRF's default exception handler to enable observing expected errors.
+
+    In addition to the default behaviour, add logging and monitoring for expected errors.
+    """
+    # Call REST framework's default exception handler first to get the standard error response.
+    response = exception_handler(exc, context)
+
+    try:
+        request = context['request'] if 'request' in context else None
+    except TypeError:  # when context is not iterable
+        request = None
+
+    _log_and_monitor_expected_errors(request, exc, 'drf')
+    return response
+
+
+class ExpectedErrorMiddleware:
+    """
+    Middleware to add logging and monitoring for expected errors.
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        response = self.get_response(request)
+        return response
+
+    def process_exception(self, request, exception):
+        """
+        Add logging and monitoring of expected errors.
+        """
+        _log_and_monitor_expected_errors(request, exception, 'middleware')
+
+
+# .. setting_name: EXPECTED_ERRORS
+# .. setting_default: None
+# .. setting_description: Used to configure logging and monitoring for expected errors.
+#     This setting is configured of a list of dicts. See setting and toggle annotations for
+#     ``EXPECTED_ERRORS[N]['XXX']`` for details of each item in the dict.
+#     If this setting is a non-empty list, all uncaught errors processed will get a ``checked_error_expected_from``
+#     attribute, whether they are expected or not. This can be used to ensure that all uncaught errors are actually
+#     processed. Those errors that are processed and match a 'MODULE_AND_CLASS' (documented elsewhere), will get an
+#     ``error_expected`` custom attribute. The value of the custom attribute will include the error class module, name
+#     and message. Unexpected errors would be errors with ``error_expected IS NULL``.
+# .. setting_warning: We use Django Middleware and a DRF custom error handler to find uncaught errors. It is still
+#     possible that some errors could slip by these mechanisms.
+
+# .. setting_name: EXPECTED_ERRORS[N]['MODULE_AND_CLASS']
+# .. setting_default: None
+# .. setting_description: Required error module and class name that is expected. For example,
+#     ``rest_framework.exceptions.PermissionDenied``.
+
+# .. toggle_name: EXPECTED_ERRORS[N]['IS_IGNORED']
+# .. toggle_implementation: DjangoSetting
+# .. toggle_default: True
```

Comment:
I was feeling a little unsure of whether this should default True or False.  Do you have thoughts on risks/benefits here?

Reply:
I defaulted to True because it would be easier to detect this problem.  We will have an alert to see if ignored errors are not being ignored, to protect against the recent issue we had with New Relic.  The runbook would include a potential misconfiguration, which might be that they wanted it ignored in New Relic but it isn't, or where they forgot to set IS_IGNORED to `False`.

If we default to `False`, or if someone mistakenly overrides to `False`, I think it will be harder to automatically detect if it is incorrectly configured.

I could add some of this to the doc if it is helpful context.

Reply:
Makes sense! Not sure what or how much of that reasoning needs to be documented -- I'll leave that up to you.

Reply:
Possible addition to the annotation:
```
Note: This is defaulted to True because it will be easier for us to detect if True is not the correct value, by seeing that these errors aren't actually ignored.
```

Reply:
Sounds great.

Reply:
Pushed.